### PR TITLE
BUD-108 Algolia Search

### PR DESCRIPTION
**Pull request by noahtallen**
Hear ye hear ye, algolia search ye

1. Installs Algolia SDK
2. Exposes a class with a function which allows you to search with any of the following optional parameters:
  -  `withText: String`
  - `matchingAnyTopicOf: [String]?`
  - `startingAt: Date?`
  - `endingAt: Date?`
  - `atLocation: (Double, Double)?` (Lat, Lng)
  - `upToDisatnce: Int = 20000` In meters, defaults to 20km
  - `usingIndex: SearchIndex` Defaults to the index for activities

If no search parameters are given, it probably returns a lot of results for the index.

Note that the completion handler gives you an array of strings, which are the activity IDs which match the search query.

It should be well tested. However, it was difficult to do "integration-type" testing, so I suppose that will come as we add in the search UI.

### CODE REVIEW

File: Buddies/Utilities/AlgoliaSearch.swift [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+//  AlgoliaSearch.swift
+//  Buddies
+//
+//  Created by Noah Allen on 2/12/19.
+//  Copyright © 2019 Jack and the Beans. All rights reserved.
+//
+
+// Usage example:
+// let formatter = DateFormatter()
+// formatter.dateFormat = "yyyy/MM/dd"
```
Comment (lukemeierj) [inline Buddies/Utilities/AlgoliaSearch.swift]: Now we can do something like`"yyyy/MM/dd".toDate()`

File: BuddiesTests/Utilities/AlgoliaTest.swift [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        let end = formatter.date(from: "2019/02/20")
+        let topics = ["a", "b"]
+        search.searchActivities(matchingAnyTopicOf: topics, startingAt: start, endingAt: end, usingIndex: index) { (ids, err) in
+            var checker = false
+            XCTAssert(ids.count > 0, "The query completed.")
+            if let query = index.query {
+                let filterText = query.filters
+                checker = filterText == "(end_time_num >= 1550206800000 AND start_time_num <= 1550638800000) AND (topic_ids:a OR topic_ids:b)"
+            }
+            XCTAssert(checker, "Constructs topic AND date query correctly.")
```
Comment (lukemeierj) [inline BuddiesTests/Utilities/AlgoliaTest.swift:64]: For my date debugging stuff, I had to bite the bullet and print the output to see what was going on in Travis.

File: Buddies/Utilities/AlgoliaSearch.swift (hunk 1 above)
Comment (noahtallen) [reply Buddies/Utilities/AlgoliaSearch.swift]: Do I need to import anything to make that happen?

File: BuddiesTests/Utilities/AlgoliaTest.swift (hunk 2 above)
Comment (noahtallen) [reply BuddiesTests/Utilities/AlgoliaTest.swift:64]: Doing that now.

File: Buddies/Utilities/AlgoliaSearch.swift (hunk 1 above)
Comment (noahtallen) [reply Buddies/Utilities/AlgoliaSearch.swift]: Got it working 👍

File: Buddies/Utilities/AlgoliaSearch.swift [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        return index(withName: indexName)
+    }
+}
+
+class AlgoliaSearch {
+    static let ACTIVITY_INDEX = "BUD_ACTIVITIES"
+    var client: SearchClient
+
+    init (algoliaClient: SearchClient? = nil) {
+        self.client = algoliaClient != nil ? algoliaClient! : AlgoliaSearch.algoliaFactory()
```
Comment (JakeThurman) [inline Buddies/Utilities/AlgoliaSearch.swift]: This could be `algoliaClient ?? AlgoliaSearch.algoliaFactory()` which I think looks a little nicer.

Not crucial.

File: Buddies/Utilities/AlgoliaSearch.swift [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    // Constructs an instance of the Algolia Client SDK:
+    static func algoliaFactory() -> Client {
+        var algoliaAppID = "NOPE"
+        var algoliaApiKey = "NOPE"
```
Comment (JakeThurman) [inline Buddies/Utilities/AlgoliaSearch.swift]: Is there any risk of error by just using "NOPE"?

File: Buddies/Utilities/AlgoliaSearch.swift [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+    // Searches for Activities with optional params
+    func searchActivities(
+              withText: String? = nil,
+              matchingAnyTopicOf: [String]? = nil,
+              startingAt: Date? = nil,
+              endingAt: Date? = nil,
+              atLocation: (Double, Double)? = nil, // Tuple: (lat, lng)
+              upToDisatnce: Int = 20000, // In meters, defaults to 20km
+              usingIndex: SearchIndex? = nil,
```
Comment (JakeThurman) [inline Buddies/Utilities/AlgoliaSearch.swift:73]: So callers need to understand indexes? Or is this just for unit testing? I assume just unit tests

File: Buddies/Utilities/AlgoliaSearch.swift [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+              completionHandler: @escaping ([String], Error?) -> Void) {
+
+        // Initialize activity index if no index is given:
+        let index = usingIndex == nil ? self.client.getIndex(withName: AlgoliaSearch.ACTIVITY_INDEX) : usingIndex!
```
Comment (JakeThurman) [inline Buddies/Utilities/AlgoliaSearch.swift]: Here's another case where the `??` could make this a little easier to read imo (and the `query` line below this)

**Review by JakeThurman — APPROVED** [review]
Other than my question, and `??` suggestion, this LGTM. Good work!

File: Buddies/Utilities/AlgoliaSearch.swift (hunk 3 above)
Comment (noahtallen) [reply Buddies/Utilities/AlgoliaSearch.swift]: Good call

File: Buddies/Utilities/AlgoliaSearch.swift (hunk 4 above)
Comment (noahtallen) [reply Buddies/Utilities/AlgoliaSearch.swift]: No - if the keys exist, it will use the keys. If the keys do not exist (which is basically impossible since it can't actually build the code without the keys file existing), it will use "NOPE" which would cause the client to not connect to a backend since four characters aren't long enough to be a key. I think that's what we want. Alternatively, I could use empty strings.

File: Buddies/Utilities/AlgoliaSearch.swift (hunk 5 above)
Comment (noahtallen) [reply Buddies/Utilities/AlgoliaSearch.swift:73]: They do not _have_ to, but they _can_. It defaults to `nil`, in which case it will construct an instance of the activity index. It doesn't default to the actual index here because the call doesn't really fit into the space that a default param would.

File: Buddies/Utilities/AlgoliaSearch.swift (hunk 6 above)
Comment (noahtallen) [reply Buddies/Utilities/AlgoliaSearch.swift]: Also good call. I wasn't aware of that syntax when I wrote this.

File: Buddies/Utilities/AlgoliaSearch.swift (hunk 4 above)
Comment (lukemeierj) [reply Buddies/Utilities/AlgoliaSearch.swift]: Maybe empty strings? Or a more informative/specific string? Looks good as is though

File: Buddies/Utilities/AlgoliaSearch.swift (hunk 5 above)
Comment (noahtallen) [reply Buddies/Utilities/AlgoliaSearch.swift:73]: And to clarify, you can't reference `self` in the default param either, which we have to do to access the client.

File: Buddies/Utilities/AlgoliaSearch.swift (hunk 4 above)
Comment (noahtallen) [reply Buddies/Utilities/AlgoliaSearch.swift]: I've changed it to be empty strings, and also re-worked the factory test to assert that the key it initializes with is not an empty string.
